db/version_set.cc

```diff
@@ -1043,7 +1044,26 @@ void LevelIterator::Seek(const Slice& target) {
   if (file_iter_.iter() != nullptr) {
     file_iter_.Seek(target);
   }
-  SkipEmptyFileForward();
+  if (SkipEmptyFileForward() && prefix_extractor_ != nullptr &&
+      file_iter_.iter() != nullptr && file_iter_.Valid()) {
+    // We've skipped the file we initially positioned to. In the prefix
+    // seek case, it is likely that the file is skipped because of
+    // prefix bloom or hash, where more keys are skipped. We then check
+    // the current key and invalidate the iterator if the prefix is
+    // already passed. This avoids LevelIterator to skip keys. A side
+    // benefit will be that it invalidate the iterator earlier so that
+    // the upper level merging iterator needs to merge fewer child
+    // iterators.
+    Slice target_user_key = ExtractUserKey(target);
+    Slice file_user_key = ExtractUserKey(file_iter_.key());
+    if (prefix_extractor_->InDomain(target_user_key) &&
+        (!prefix_extractor_->InDomain(file_user_key) ||
+         user_comparator_.Compare(
+             prefix_extractor_->Transform(target_user_key),
+             prefix_extractor_->Transform(file_user_key)) != 0)) {
```

Comment:
It is not clear to me that why we should invalidate the iterator if the "first key" of the file has a different prefix that the target key. Could not the key after that have the same prefix as our target key?

Reply:
We don't have to invalidate the iterator, but we can. The contract of prefix iterating after the prefix allows both cases. Is your question about why we want to pick the former case here?

Reply:
"The contract of prefix iterating after the prefix allows both cases." I think this would be a helpful sentence in inline comments. I am thinking of a future refactorer trying to figure how essential this step. Saying that the contract allow both but we change it to do invalidate, and mentioning the benefit would help the future refactorer.
Currently it only says "This avoids LevelIterator to skip keys. " which is quite ambiguous to me. The it talks about a side-benefit, which sounds like an optimization. If that is the case, it is better to be mentioned.

What I understood so far from your comments is this: "The contract allow two implementations: to invalidate or not to invalidate. We change the implementation to "to invalidate" to optimize for performance.

Reply:
Does this require using user_comparator for comparison? I think using Slice's own compare method is sufficient. Using the former to compare prefixes might result in undefined behavior.